File: firmware/components/cortex/src/platform/stomach_cp_cache.rs
use core::mem::MaybeUninit;

use crate::ffi::stomach_cp_cache;

pub use crate::ffi::stomach_cp_cache::{
    StomachCpCacheRaw, STOMACH_CP_CACHE_SIZE, STOMACH_CP_DIGEST_LEN,
};

pub struct StomachCpCache {
    raw: StomachCpCacheRaw,
}

impl StomachCpCache {
    pub const CAPACITY: usize = STOMACH_CP_CACHE_SIZE;

    pub fn new() -> Self {
        Self { raw: new_raw() }
    }

    pub fn clear(&mut self) {
        clear(&mut self.raw);
    }

    pub fn contains(&self, digest: &[u8; STOMACH_CP_DIGEST_LEN]) -> bool {
        contains(&self.raw, digest)
    }

    pub fn seen_or_add(&mut self, digest: &[u8; STOMACH_CP_DIGEST_LEN]) -> bool {
        seen_or_add(&mut self.raw, digest)
    }

    pub fn count(&self) -> usize {
        count(&self.raw)
    }

    pub fn as_raw(&self) -> &StomachCpCacheRaw {
        &self.raw
    }

    pub fn as_raw_mut(&mut self) -> &mut StomachCpCacheRaw {
        &mut self.raw
    }
}

impl Default for StomachCpCache {
    fn default() -> Self {
        Self::new()
    }
}

pub fn new_raw() -> StomachCpCacheRaw {
    let mut cache = MaybeUninit::<StomachCpCacheRaw>::uninit();
    unsafe {
        stomach_cp_cache::stomach_cp_cache_init(cache.as_mut_ptr());
        cache.assume_init()
    }
}

pub fn init(cache: &mut StomachCpCacheRaw) {
    unsafe { stomach_cp_cache::stomach_cp_cache_init(cache) };
}

pub fn clear(cache: &mut StomachCpCacheRaw) {
    unsafe { stomach_cp_cache::stomach_cp_cache_clear(cache) };
}

pub fn contains(cache: &StomachCpCacheRaw, digest: &[u8; STOMACH_CP_DIGEST_LEN]) -> bool {
    unsafe { stomach_cp_cache::stomach_cp_cache_contains(cache, digest.as_ptr()) }
}

pub fn seen_or_add(cache: &mut StomachCpCacheRaw, digest: &[u8; STOMACH_CP_DIGEST_LEN]) -> bool {
    unsafe { stomach_cp_cache::stomach_cp_cache_seen_or_add(cache, digest.as_ptr()) }
}

pub fn count(cache: &StomachCpCacheRaw) -> usize {
    unsafe { stomach_cp_cache::stomach_cp_cache_count(cache) }
}

File: firmware/components/stomach/stomach_cp_cache.c
#include "stomach_cp_cache.h"

#include <string.h>

#include "mbedtls/platform_util.h"

static bool digest_equals(
    const uint8_t left[STOMACH_CP_DIGEST_LEN],
    const uint8_t right[STOMACH_CP_DIGEST_LEN]
) {
    return memcmp(left, right, STOMACH_CP_DIGEST_LEN) == 0;
}

void stomach_cp_cache_init(stomach_cp_cache_t *cache) {
    stomach_cp_cache_clear(cache);
}

void stomach_cp_cache_clear(stomach_cp_cache_t *cache) {
    if (!cache) {
        return;
    }

    mbedtls_platform_zeroize(cache->entries, sizeof(cache->entries));
    cache->next = 0;
    cache->count = 0;
}

bool stomach_cp_cache_contains(
    const stomach_cp_cache_t *cache,
    const uint8_t digest[STOMACH_CP_DIGEST_LEN]
) {
    if (!cache || !digest) {
        return false;
    }

    for (size_t i = 0; i < cache->count; ++i) {
        if (digest_equals(cache->entries[i], digest)) {
            return true;
        }
    }

    return false;
}

bool stomach_cp_cache_seen_or_add(
    stomach_cp_cache_t *cache,
    const uint8_t digest[STOMACH_CP_DIGEST_LEN]
) {
    if (!cache || !digest) {
        return true;
    }

    if (stomach_cp_cache_contains(cache, digest)) {
        return true;
    }

    memcpy(cache->entries[cache->next], digest, STOMACH_CP_DIGEST_LEN);
    cache->next = (cache->next + 1) % STOMACH_CP_CACHE_SIZE;

    if (cache->count < STOMACH_CP_CACHE_SIZE) {
        ++cache->count;
    }

    return false;
}

size_t stomach_cp_cache_count(const stomach_cp_cache_t *cache) {
    return cache ? cache->count : 0;
}
